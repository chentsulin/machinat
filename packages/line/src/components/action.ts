import { partSegment } from '@machinat/core/renderer';
import { PartSegment } from '@machinat/core/renderer/types';
import { annotateLineComponent } from '../utils';
import type { LineComponent } from '../types';

/**
 * @category Props
 */
type PostbackActionProps = {
  /** Label for the action */
  label: string;
  /**
   * String returned via webhook in the `postback.data` property of the postback
   * event. Max character limit: 300.
   */
  data: string;
  /**
   * Text displayed in the chat as a message sent by the user when the action is
   * performed. Required for quick reply buttons. Optional for the other message
   * types. Max character limit: 300.
   */
  displayText?: string;
  /** Alias of `displayText` */
  text?: string;
};

/** @internal */
const __PostbackAction = function PostbackAction(node, path) {
  const { label, data, text, displayText } = node.props;
  return [
    partSegment(node, path, {
      type: 'postback',
      data,
      label,
      displayText: displayText || text,
    }),
  ];
};
/**
 * When a control associated with this action is tapped, a postback event is
 * returned via webhook with the specified string in the data property.
 * @category Component
 * @props {@link PostbackActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#postback-action).
 */
export const PostbackAction: LineComponent<
  PostbackActionProps,
  PartSegment<any>
> = annotateLineComponent(__PostbackAction);

/**
 * @category Props
 */
type MessageActionProps = {
  /** Label for the action */
  label: string;
  /** Text sent when the action is performed. Max character limit: 300 */
  text?: string;
};

/** @internal */
const __MessageAction = function MessageAction(node, path) {
  const { label, text } = node.props;
  return [
    partSegment(node, path, {
      type: 'message',
      label,
      text,
    }),
  ];
};
/**
 * When a control associated with this action is tapped, the string in the
 * `text` property is sent as a message from the user.
 * @category Component
 * @props {@link MessageActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#message-action).
 */
export const MessageAction: LineComponent<
  MessageActionProps,
  PartSegment<any>
> = annotateLineComponent(__MessageAction);

/**
 * @category Props
 */
type UriActionProps = {
  /** Label for the action */
  label: string;
  /**
   * URI opened when the action is performed (Max character limit: 1000). The
   * available schemes are http, https, line, and tel.
   */
  uri: string;
};

/** @internal */
const __UriAction = function UriAction(node, path) {
  const { label, uri } = node.props;
  return [
    partSegment(node, path, {
      type: 'uri',
      label,
      uri,
    }),
  ];
};
/**
 * When a control associated with this action is tapped, the URI specified in
 * the `uri` property is opened.
 * @category Component
 * @props {@link UriActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#uri-action).
 */
export const UriAction: LineComponent<
  UriActionProps,
  PartSegment<any>
> = annotateLineComponent(__UriAction);

/** @internal */
const pad2 = (n) => (n < 10 ? `0${n}` : n);

/** @internal */
const fullDate = (d) =>
  `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}`;

/** @internal */
const fullHourMinute = (d) => `${pad2(d.getHours())}:${pad2(d.getMinutes())}`;

/** @internal */
const dateToStringByMode = (mode, d) =>
  !(d instanceof Date)
    ? d
    : mode === 'datetime'
    ? `${fullDate(d)}T${fullHourMinute(d)}`
    : mode === 'date'
    ? fullDate(d)
    : fullHourMinute(d);

/**
 * @category Props
 */
type DateTimePickerActionProps = {
  /** Label for the action  */
  label: string;
  /**
   * String returned via webhook in the postback.data property of the postback
   * event. Max character limit: 300.
   */
  data: string;
  /** Action mode */
  mode: 'datetime' | 'date' | 'time';
  /** Initial value of date or time */
  initial?: string | Date;
  /**
   * Largest date or time value that can be selected. Must be greater than the
   * min value.
   */
  max?: string | Date;
  /**
   * Smallest date or time value that can be selected. Must be less than the max
   * value.
   */
  min?: string | Date;
};

/** @internal */
const __DateTimePickerAction = function DateTimePickerAction(node, path) {
  const { label, data, mode = 'datetime', initial, min, max } = node.props;

  return [
    partSegment(node, path, {
      type: 'datetimepicker',
      label,
      data,
      mode,
      initial: dateToStringByMode(mode, initial),
      max: dateToStringByMode(mode, max),
      min: dateToStringByMode(mode, min),
    }),
  ];
};
/**
 * When a control associated with this action is tapped, a postback event is
 * returned via webhook with the date and time selected by the user from the
 * date and time selection dialog. The datetime picker action does not support
 * time zones.
 * @category Component
 * @props {@link DateTimePickerActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#datetime-picker-action).
 */
export const DateTimePickerAction: LineComponent<
  DateTimePickerActionProps,
  PartSegment<any>
> = annotateLineComponent(__DateTimePickerAction);

/**
 * @category Props
 */
type CameraActionProps = {
  /** Label for the action. Max character limit: 20 */
  label: string;
};

/** @internal */
const __CameraAction = function CameraAction(node, path) {
  return [
    partSegment(node, path, {
      type: 'camera',
      label: node.props.label,
    }),
  ];
};
/**
 * This action can be configured only with quick reply buttons. When a button
 * associated with this action is tapped, the camera screen in LINE is opened.
 * @category Component
 * @props {@link CameraActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#camera-action).
 */
export const CameraAction: LineComponent<
  CameraActionProps,
  PartSegment<any>
> = annotateLineComponent(__CameraAction);

/**
 * @category Props
 */
type CameraRollActionProps = {
  /** Label for the action. Max character limit: 20 */
  label: string;
};

/** @internal */
const __CameraRollAction = function CameraRollAction(node, path) {
  return [
    partSegment(node, path, {
      type: 'cameraRoll',
      label: node.props.label,
    }),
  ];
};
/**
 * This action can be configured only with quick reply buttons. When a button
 * associated with this action is tapped, the camera roll screen in LINE is
 * opened.
 * @category Component
 * @props {@link CameraRollActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#camera-roll-action).
 */
export const CameraRollAction: LineComponent<
  CameraRollActionProps,
  PartSegment<any>
> = annotateLineComponent(__CameraRollAction);

/**
 * @category Props
 */
type LocationActionProps = {
  /** Label for the action. Max character limit: 20 */
  label: string;
};

/** @internal */
const __LocationAction = function LocationAction(node, path) {
  return [
    partSegment(node, path, {
      type: 'location',
      label: node.props.label,
    }),
  ];
};
/**
 * This action can be configured only with quick reply buttons. When a button
 * associated with this action is tapped, the location screen in LINE is opened.
 * @category Component
 * @props {@link LocationActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#camera-roll-action).
 */
export const LocationAction: LineComponent<
  LocationActionProps,
  PartSegment<any>
> = annotateLineComponent(__LocationAction);

export type Action =
  | typeof PostbackAction
  | typeof MessageAction
  | typeof UriAction
  | typeof DateTimePickerAction
  | typeof CameraAction
  | typeof CameraRollAction
  | typeof LocationAction;
