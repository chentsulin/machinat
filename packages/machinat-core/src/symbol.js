// @flow
export const MACHINAT_ELEMENT_TYPE = Symbol('machinat.element');

export const MACHINAT_FRAGMENT_TYPE = Symbol('machinat.element.fragment');

export const MACHINAT_PAUSE_TYPE = Symbol('machinat.element.pause');

export const MACHINAT_PROVIDER_TYPE = Symbol('machinat.element.provider');

export const MACHINAT_THUNK_TYPE = Symbol('machinat.element.thunk');

export const MACHINAT_RAW_TYPE = Symbol('machinat.element.raw');

export const MACHINAT_NATIVE_TYPE = Symbol('machinat.component.native');

export const MACHINAT_SERVICES_CONTAINER = Symbol(
  'machinat.services.container'
);

export const MACHINAT_SERVICES_INTERFACE = Symbol(
  'machinat.services.interface'
);

export const MACHINAT_SERVICES_PROVIDER = Symbol('machinat.services.provider');
