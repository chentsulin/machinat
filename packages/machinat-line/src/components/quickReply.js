/* eslint-disable import/prefer-default-export */
import { partSegment } from '@machinat/core/renderer';
import { annotateLineComponent } from '../utils';

export const QuickReply = async (node, path, render) => {
  const { imageURL, action } = node.props;

  const actionSegments = await render(action, '.action');
  const actionValue = actionSegments?.[0].value;

  return [
    partSegment(node, path, {
      type: 'action',
      imageUrl: imageURL,
      action: actionValue,
    }),
  ];
};

annotateLineComponent(QuickReply);
