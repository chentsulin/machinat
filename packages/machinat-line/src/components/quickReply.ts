/* eslint-disable import/prefer-default-export */
import { partSegment } from '@machinat/core/renderer';
import { annotateLineComponent } from '../utils';

export const QuickReply = async (node, path, render) => {
  const { imageUrl, imageURL, action } = node.props;

  const actionSegments = await render(action, '.action');
  const actionValue = actionSegments?.[0].value;

  return [
    partSegment(node, path, {
      type: 'action',
      imageUrl: imageUrl || imageURL,
      action: actionValue,
    }),
  ];
};

annotateLineComponent(QuickReply);
