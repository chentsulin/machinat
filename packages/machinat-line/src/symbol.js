export const LINE_NAITVE_TYPE = Symbol.for('machinat.line.native');

export const API_PATH = Symbol.for('machinat.line.api_path');
export const HAS_BODY = Symbol.for('machinat.line.has_body');
export const THREAD_ID = Symbol.for('machinat.line.thread_id');

export const NO_BODY = Symbol.for('machinat.line.act.no_body');
