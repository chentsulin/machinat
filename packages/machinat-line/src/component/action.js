import { annotate, asNative, asUnit } from 'machinat-utility';
import { LINE_NAITVE_TYPE } from '../symbol';

export const PostbackAction = ({ label, data, text, displayText }) => [
  {
    type: 'postback',
    data,
    label,
    displayText: displayText || text,
  },
];

annotate(asNative(LINE_NAITVE_TYPE), asUnit(false))(PostbackAction);

export const MessageAction = ({ label, text }) => [
  {
    type: 'message',
    label,
    text,
  },
];

annotate(asNative(LINE_NAITVE_TYPE), asUnit(false))(MessageAction);

export const URIAction = ({ label, uri }) => [
  {
    type: 'uri',
    label,
    uri,
  },
];

annotate(asNative(LINE_NAITVE_TYPE), asUnit(false))(URIAction);

const pad2 = n => (n < 10 ? `0${n}` : n);
const fullDate = d =>
  `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}`;
const fullHourMinute = d => `${pad2(d.getHours())}:${pad2(d.getMinutes())}`;
const dateToStringByMode = (mode, d) =>
  !(d instanceof Date)
    ? d
    : mode === 'datetime'
    ? `${fullDate(d)}T${fullHourMinute(d)}`
    : mode === 'date'
    ? fullDate(d)
    : fullHourMinute(d);

export const DateTimePickerAction = ({
  label,
  data,
  mode,
  date,
  time,
  initial,
  min,
  max,
}) => {
  const modeRefined =
    mode || (!date === !time ? 'datetime' : date ? 'date' : 'time');

  return [
    {
      type: 'datetimepicker',
      label,
      data,
      mode: modeRefined,
      initial: dateToStringByMode(modeRefined, initial),
      max: dateToStringByMode(modeRefined, max),
      min: dateToStringByMode(modeRefined, min),
    },
  ];
};
annotate(asNative(LINE_NAITVE_TYPE), asUnit(false))(DateTimePickerAction);

export const CameraAction = ({ label }) => [
  {
    type: 'camera',
    label,
  },
];

annotate(asNative(LINE_NAITVE_TYPE), asUnit(false))(CameraAction);

export const CameraRollAction = ({ label }) => [
  {
    type: 'cameraRoll',
    label,
  },
];

annotate(asNative(LINE_NAITVE_TYPE), asUnit(false))(CameraRollAction);

export const LocationAction = ({ label }) => [
  {
    type: 'location',
    label,
  },
];

annotate(asNative(LINE_NAITVE_TYPE), asUnit(false))(LocationAction);
