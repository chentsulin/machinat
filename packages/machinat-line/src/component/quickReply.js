/* eslint-disable import/prefer-default-export */
import {
  annotate,
  asNative,
  asUnit,
  valuesOfAssertedType,
} from 'machinat-utility';

import { LINE_NAITVE_TYPE } from '../symbol';

import * as _actionModule from './action';

const actionComponents = Object.values(_actionModule);
const renderActionValues = valuesOfAssertedType(...actionComponents);

export const QuickReply = ({ imageURL, action }, render) => {
  const actionValues = renderActionValues(action, render, 'action');

  return [
    {
      type: 'action',
      imageUrl: imageURL,
      action: actionValues && actionValues[0],
    },
  ];
};

annotate(asNative(LINE_NAITVE_TYPE), asUnit(false))(QuickReply);
