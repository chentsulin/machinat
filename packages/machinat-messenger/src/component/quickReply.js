import { annotate, asNative, asUnit } from 'machinat-utility';

import { MESSENGER_NAITVE_TYPE } from '../symbol';

export const QuickReply = ({ title, payload, imageURL }) => [
  {
    content_type: 'text',
    title,
    payload,
    image_url: imageURL,
  },
];

annotate(asNative(MESSENGER_NAITVE_TYPE), asUnit(false))(QuickReply);

const LOCATION_QUICK_REPLY_RENDERED = [{ content_type: 'location' }];
const PHONE_QUICK_REPLY_RENDERED = [{ content_type: 'user_phone_number' }];
const EMAIL_QUICK_REPLY_RENDERED = [{ content_type: 'user_email' }];

export const LocationQuickReply = () => LOCATION_QUICK_REPLY_RENDERED;

annotate(asNative(MESSENGER_NAITVE_TYPE), asUnit(false))(LocationQuickReply);

export const PhoneQuickReply = () => PHONE_QUICK_REPLY_RENDERED;

annotate(asNative(MESSENGER_NAITVE_TYPE), asUnit(false))(PhoneQuickReply);

export const EmailQuickReply = () => EMAIL_QUICK_REPLY_RENDERED;

annotate(asNative(MESSENGER_NAITVE_TYPE), asUnit(false))(EmailQuickReply);
