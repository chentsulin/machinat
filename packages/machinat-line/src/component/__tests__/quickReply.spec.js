import Machinat from 'machinat';

import { QuickReply } from '../quickReply';
import { URIAction } from '../action';

import { LINE_NAITVE_TYPE } from '../../symbol';

import render from './render';

it('is valid native component', () => {
  expect(typeof QuickReply).toBe('function');

  expect(QuickReply.$$native).toBe(LINE_NAITVE_TYPE);
  expect(QuickReply.$$entry).toBe(undefined);
  expect(QuickReply.$$unit).toBe(false);
});

it('renders match snapshot', () => {
  expect(
    render(
      <QuickReply
        imageURL="https://..."
        action={<URIAction uri="https://..." label="foo" />}
      />
    ).map(act => act.value)
  ).toMatchSnapshot();
});
