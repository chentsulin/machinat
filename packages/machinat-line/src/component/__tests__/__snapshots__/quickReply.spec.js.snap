// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders match snapshot 1`] = `
Array [
  Object {
    "action": Object {
      "label": "foo",
      "type": "uri",
      "uri": "https://...",
    },
    "imageUrl": "https://...",
    "type": "action",
  },
]
`;
